tests/RobotFramework/tests/tedge_to_te_converter/convert_tedge_topics_to_te_topics.robot

```diff
@@ -30,12 +30,44 @@ Convert child device event topic
 
 Convert main device alarm topic
     Execute Command    tedge mqtt pub tedge/alarms/minor/test_alarm '{"text":"test alarm"}' -q 2 -r
-    Should Have MQTT Messages    te/device/main///a/test_alarm    message_contains=minor
+    ${messages}=    Should Have MQTT Messages    te/device/main///a/test_alarm    minimum=1    maximum=2
+    Should Contain    ${messages[0]}    "severity":"minor"
+
+Convert main device alarm topic and retain
+    Execute Command    tedge mqtt pub tedge/alarms/minor/test_alarm '{"text":"test alarm"}' -q 2 -r
+    ${messages}=    Should Have MQTT Messages    te/device/main///a/test_alarm    minimum=1    maximum=2
+    Should Contain    ${messages[0]}    "severity":"minor"
+    # Check if the retained message received with new client or not
+    ${result}=    Execute Command    tedge mqtt sub te/device/main///a/test_alarm & sleep 2s; kill $!   
+    Should Contain    ${result}    "severity":"minor"         
 
 Convert child device alarm topic
     Execute Command    tedge mqtt pub tedge/alarms/major/test_alarm/child '{"text":"test alarm"}' -q 2 -r
-    Should Have MQTT Messages    te/device/child///a/test_alarm    message_contains=major
+    ${messages}=    Should Have MQTT Messages    te/device/child///a/test_alarm    minimum=1    maximum=2
+    Should Contain    ${messages[0]}    "severity":"major"        
+   
+
+Convert main device service health topic
+    Execute Command    tedge mqtt pub tedge/health/main-service '{"pid":1234,"status":"up"}' -q 2 -r
+    ${messages}=    Should Have MQTT Messages    te/device/main/service/main-service/status/health    minimum=1    maximum=2
+    Should Contain    ${messages[0]}    "pid":1234
+    Should Contain    ${messages[0]}    "status":"up"    
```

Comment:
Since we know the message will be returned as json, then we can use another robot framework (the dependency is already there) to parse the response and apply nicer assertions (that don't depend on the format of the json, e.g. is pretty printed or not).

To use `JSONLibrary` you need to add the import to the `*** Settings ***` section of the file (at the top of the file).

```
Library             JSONLibrary
```

Then you can rewrite the tests as follows:

```
Convert main device service health topic
    Execute Command    tedge mqtt pub tedge/health/main-service '{"pid":1234,"status":"up"}' -q 2 -r
    ${messages}=    Should Have MQTT Messages    te/device/main/service/main-service/status/health    minimum=1    maximum=2
    ${message}=    Convert String To Json    ${messages[0]}
    Should Be Equal As Numbers    ${message["pid"]}    1234
    Should Be Equal    ${message["status"]}    up
```

This same approach can be applied to all test cases that you added.

Reply:
Updated